TYP: avoid inherit_names for DatetimeIndexOpsMixin

twoertwein opened the pull request:
Similar to #36742
 
xref #32100

twoertwein committed:
TYP: avoid inherit_names for DatetimeIndexOpsMixin

twoertwein commented:
Over multiple PRs, my goal is to remove `inherit_names` to make static type checkers aware of those methods.

The alternative is to keep `inherit_names` but add `method_name: type_annotation`. Writing those annotations is tricky for properties.

Adding little helper methods as done in #36742 and here has the advantage that we can easily express the type of properties, it is more difficult for the implementation and the annotations to get out of sync, and it might also be more readable. Mypy doesn't seem to support properties with decorators but it seems that it just ignores them, which should be fine in most/all cases (`@doc(...)` doesn't change the function signature)

jbrockmendel commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -214,7 +248,8 @@ def _summary(self, name=None) -> str:
         # e.g. test_getitem_setitem_periodindex
         # History of conversation GH#3452, GH#3931, GH#2369, GH#14826
-        return reso > self._resolution_obj
+        # error: Unsupported operand types for < ("str" and "Resolution")
```
it thinks that reso is a str?

twoertwein replied:
must have copied the wrong error message. The "correct" error is

> error: Unsupported left operand type for > ("Resolution")  [operator] 
> note: Right operand is of type "Optional[Resolution]"

twoertwein replied:
I added a breakpoint, both of them are Resolution objects

mroeschke reviewed (approved):
LGTM cc @jbrockmendel

jbrockmendel commented on pandas/core/indexes/period.py:
```diff
@@ -392,7 +391,8 @@ def is_full(self) -> bool:
             raise ValueError("Index is not monotonic")
         values = self.asi8
-        return ((values[1:] - values[:-1]) < 2).all()
+        # error: Incompatible return value type (got "bool_", expected "bool")
```
worth wrapping in `bool(...)`?

jbrockmendel commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -88,9 +80,51 @@ class DatetimeIndexOpsMixin(NDArrayBackedExtensionIndex):
+    # error: Decorated property not supported
+    @cache_readonly  # type: ignore[misc]
+    @doc(DatetimeLikeArrayMixin.inferred_freq)
+    def inferred_freq(self) -> str | None:
```
i think this is only for DTI/TDI, not PI?

twoertwein replied:
On 1.4.3 (and on this PR) it errors for PeriodIndex:

```py
>>> pd.PeriodIndex(year=[2000, 2002], quarter=[1, 3]).inferred_freq
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/properties.pyx", line 37, in pandas._libs.properties.CachedProperty.__get__
  File "pandas/core/indexes/extension.py", line 59, in cached
    return getattr(self._data, name)
  File "pandas/core/arrays/datetimelike.py", line 917, in inferred_freq
    return frequencies.infer_freq(self)
  File "pandas/tseries/frequencies.py", line 183, in infer_freq
    raise TypeError(
TypeError: PeriodIndex given. Check the `freq` attribute instead of using infer_freq.
```

Could add a `NoReturn`-annotation to `freqstr` in `PeriodIndex`.

jbrockmendel replied:
yah, so better to just move this to DatetimeTimedeltaMixin?

twoertwein replied:
Thank you! I moved it.

jbrockmendel commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -88,9 +80,51 @@ class DatetimeIndexOpsMixin(NDArrayBackedExtensionIndex):
+        return self._data.inferred_freq
+
+    @cache_readonly
+    def _resolution_obj(self) -> Resolution | None:
```
is None possible?

twoertwein replied:
`DatetimeLikeArrayMixin._resolution_obj` can theoretically be `None` but that is never the case when running the tests!

I could make sense to simplify `DatetimeLikeArrayMixin._resolution_obj` https://github.com/pandas-dev/pandas/blob/926b9ceff10d9b7a957811f0a4de3167332196de/pandas/core/arrays/datetimelike.py#L968 to

```py
    @property
    def _resolution_obj(self) -> Resolution:
        freqstr = self.freqstr
        return Resolution.get_reso_from_freqstr(freqstr)  # the KeyError is never triggered in the tests
```

twoertwein replied:
mh, in test pandas/tests/arrays/test_datetimes.py(82)test_fields(), `self.freq` is None but `_resolution_obj` is not called during this test but it would return `None` in this case. I came across this when testing whether `freqstr` can be `None` (and in this case it is `None`): if `freqstr` can be `None`, `_resolution_obj` can be `None`.

One argument for removing `None` from at least `_resolution_obj` (but not `freqstr`) is that places that call it, do not check for None and would anyways error there

jbrockmendel replied:
Yah, looks like it is only relevant for TDA, and we could re-use the logic (minus self.tz) from DTA._resolution_obj or that

twoertwein replied:
Sorry, I'm not following: do you suggest moving `_resolution_obj` (or `freqstr`) to a different class?

twoertwein replied:
`DatetimeLikeArrayMixin._resolution_obj` is only ever called for `PeriodArray` in the tests (other arrays might implement their own version?) and `DatetimeIndexOpsMixin._resolution_obj` is never called at least in the tests (I guess sub-classes implement it).

Cannot make `DatetimeIndexOpsMixin._resolution_obj` abstract as `DatetimeIndex` and `TimedeltaIndex` do not "implement" it (`DatetimeIndex` uses `inherit_names` to replace it but it didn't see anything for `TimedeltaIndex`).

twoertwein replied:
I would be fine leaving it as `Resolution | None` for now.

jbrockmendel replied:
if we don't need TDA._resolution_obj to exist, then we can just move it.  otherwise, i think something like

```
@property
def _resolution_obj(self) -> Resolution:
     tz = getattr(self, "tz", None)  # i.e. None for TimedeltaArray
    return get_resolution(self.asi8, tz, reso=self._reso)
```

works for DTA/TDA, and the current DatetimeLikeArrayMixin._resolution_obj becomes PeriodArray._resolution_obj

twoertwein replied:
Is `TimedeltaIndex` supposed to have `_resolution_obj` (and `resolution`):

```
>>> pd.TimedeltaIndex([1])._resolution_obj  # None
>>> pd.TimedeltaIndex([1]).resolution
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/properties.pyx", line 37, in pandas._libs.properties.CachedProperty.__get__
  File "python3.10/site-packages/pandas/core/indexes/extension.py", line 59, in cached
    return getattr(self._data, name)
  File "python3.10/site-packages/pandas/core/arrays/datetimelike.py", line 937, in resolution
    return self._resolution_obj.attrname  # type: ignore[union-attr]
AttributeError: 'NoneType' object has no attribute 'attrname'
```

If it is not supposed to have that, a good solution (for at least for the index classes) would be to make `DatetimeIndexOpsMixin._resolution_obj -> Resolution` abstract (all sub-classes except TimedeltaIndex implement it and return `Resolution`).

jbrockmendel commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -88,9 +80,51 @@ class DatetimeIndexOpsMixin(NDArrayBackedExtensionIndex):
+        return self._data.freq
+
+    @freq.setter
+    def freq(self, value) -> None:
```
do we allow setting?

twoertwein replied:
It works at least on main `pd.TimedeltaIndex([]).freq = None` but I don't know whether that is meant to be the case.

Happy to remove the setter (seems like that would be something that is not possible with `inherit_names`: it would copy both the getter and setter),

twoertwein replied:
It cannot be set for `PeriodArray` but other arrays (and index) types support it.

pandas/tests/arrays/categorical/test_constructors.py has two tests that set freq to None (but the tests also pass when removing those lines).

twoertwein replied:
I moved it to `DatetimeTimedeltaMixin`

jbrockmendel commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -926,7 +927,9 @@ def sequence_to_td64ns(
     assert data.dtype == "m8[ns]", data
-    return data, inferred_freq
+    # error: Incompatible return value type (got "Tuple[Any, Optional[BaseOffset]]",
+    # expected "Tuple[ndarray[Any, Any], Optional[Tick]]")
```
looks like it doesn't recognize that `data` is an ndarray here?

jbrockmendel commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -446,7 +446,8 @@ def __truediv__(self, other):
             if self.freq is not None:
                 # Tick division is not implemented, so operate on Timedelta
-                freq = self.freq.delta / other
+                # error: "BaseOffset" has no attribute "delta"
```
TimedeltaArray.freq should be `Tick|None`

jbrockmendel commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1526,8 +1539,9 @@ def __sub__(self, other):
         elif is_integer_dtype(other_dtype):
             if not is_period_dtype(self.dtype):
                 raise integer_op_not_supported(self)
+            # error: Item "None" of "Optional[BaseOffset]" has no attribute "n"
```
PeriodArray.freq should be non-None

jbrockmendel commented:
Not a hill I plan to die on, but it isn't obvious to me that this is worthwhile.  we implemented inherit_names in the first place for boilerplate-reduction, which is still a good goal.

twoertwein committed:
address some of the comments

twoertwein committed:
move inferred_freq

twoertwein committed:
move freq-setter in DatetimeTimedeltaMixin

twoertwein committed:
Merge branch 'main' into datetimelike

twoertwein committed:
Merge branch 'main' into datetimelike

twoertwein committed:
fix dev test

twoertwein committed:
Merge branch 'main' into datetimelike

twoertwein committed:
make _resolution_obj abstract

twoertwein committed:
use property in abstract class

twoertwein committed:
fix order

twoertwein commented:
I think I addressed all comments but now we also have some small implementation changes. If you could please have another look @jbrockmendel  @mroeschke

mroeschke commented:
Generally looks okay to me. I don't have a strong opinion in which direction `inherit_names` should go (but generally I don't mind some boilerplating as it makes the code easier to grok IMO)

jbrockmendel commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -552,7 +552,12 @@ def _concat_same_type(
                 pairs = zip(to_concat[:-1], to_concat[1:])
-                if all(pair[0][-1] + obj.freq == pair[1][0] for pair in pairs):
+                # error: No overload variant of "__radd__" of "BaseOffset" matches
+                # argument type "NaTType"
```
is this something that should be fixed in the BaseOffset .pyi file?  (OK if you want to leave that as out of scope)

twoertwein replied:
I added `NaTType` for just `__radd__`. Would need to add it to many more oprators.

jbrockmendel commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1220,7 +1225,10 @@ def _sub_period(self, other: Period) -> npt.NDArray[np.object_]:
             self.asi8, -other.ordinal, arr_mask=self._isnan
         )
-        new_data = np.array([self.freq.base * x for x in new_i8_data])
+        # error: Item "None" of "Optional[BaseOffset]" has no attribute "base"
```
can do `self = cast("PeriodArray", self)` above

jbrockmendel commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1427,8 +1435,10 @@ def _time_shift(
 
-        start = self[0] + periods * self.freq
-        end = self[-1] + periods * self.freq
+        # error: No overload variant of "__radd__" of "BaseOffset" matches
```
not sure if there's any way to annotate this, but the fact that we excluded `self.freq is None` above means self[0] and self[-1] can't be NaT

twoertwein replied:
this is now anyways covered by the new NaTType overload in offsets.pyi :)

jbrockmendel commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1456,8 +1466,9 @@ def __add__(self, other):
             # as is_integer returns True for these
             if not is_period_dtype(self.dtype):
                 raise integer_op_not_supported(self)
+            # error: Item "None" of "Optional[BaseOffset]" has no attribute "n"
```
i think need to do the same cast on the self in self.freq.n?

jbrockmendel commented on pandas/core/arrays/timedeltas.py:
```diff
@@ -143,6 +144,23 @@ def _box_func(self, x: np.timedelta64) -> Timedelta | NaTType:
+    # error: Decorated property not supported
+    @property  # type: ignore[misc]
+    @doc(dtl.DatetimeLikeArrayMixin.freq)
+    def freq(self) -> Tick:
```
TDA.freq _can_ be None.  its only PeriodArray.freq that can't be None

twoertwein replied:
Changed :)

jbrockmendel commented on pandas/core/indexes/period.py:
```diff
@@ -393,7 +392,7 @@ def is_full(self) -> bool:
             raise ValueError("Index is not monotonic")
         values = self.asi8
-        return ((values[1:] - values[:-1]) < 2).all()
+        return ((values[1:] - values[:-1]) < 2).all().item()
```
why is this needed?  seems like the kind of thing that might be fragile across numpy versions

twoertwein replied:
The function is annotated as returning bool, but it currently return numpy.bool_ (which does not seem to inherit from bool):

```py
>>> type((np.random.rand(3) > 0.5).all())
<class 'numpy.bool_'>
>>> isinstance((np.random.rand(3) > 0.5).all(), bool)
False
```

twoertwein committed:
address non-cast comments

twoertwein committed:
add casts

twoertwein committed:
Merge branch 'main' into datetimelike

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

twoertwein commented:
> Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

Opened #49804
